From 0427fdfe2b0f302b7ebd248f36c2c9b6c0eeba19 Mon Sep 17 00:00:00 2001
From: Quintus Chu <@gmail.com>
Date: Tue, 29 Dec 2020 04:58:05 +0800
Subject: [PATCH]Add filter-aaaa option at dnsmasq for the FriendlyARM NanoPi R4S.

Signed-off-by: Quintus Chu
Signed-off-by: Chuck <fanck0605@qq.com>
---

            new file:   package/network/services/dnsmasq/patches/900-add-filter-aaaa-option.patch
            modified:   feeds/luci/modules/luci-mod-network/htdocs/luci-static/resources/view/network/dhcp.js
            modified:   package/base-files/files/etc/init.d/boot
            modified:   package/network/services/dnsmasq/files/dhcp.conf
            modified:   package/network/services/dnsmasq/files/dnsmasq.init

diff --git a/feeds/luci/modules/luci-mod-network/htdocs/luci-static/resources/view/network/dhcp.js b/feeds/luci/modules/luci-mod-network/htdocs/luci-static/resources/view/network/dhcp.js
index 541c2fca1b..6aae03e54b 100644
--- a/feeds/luci/modules/luci-mod-network/htdocs/luci-static/resources/view/network/dhcp.js
+++ b/feeds/luci/modules/luci-mod-network/htdocs/luci-static/resources/view/network/dhcp.js
@@ -199,6 +199,11 @@ return view.extend({
 		s.taboption('files', form.DynamicList, 'addnhosts',
 			_('Additional Hosts files')).optional = true;
 
+		o = s.taboption('advanced', form.Flag, 'filteraaaa',
+			_('Filter IPv6 Records'),
+			_('Filter IPv6(AAAA) Records during DNS resolution'));
+		o.optional = true;
+
 		o = s.taboption('advanced', form.Flag, 'quietdhcp',
 			_('Suppress logging'),
 			_('Suppress logging of the routine operation of these protocols'));
diff --git a/package/base-files/files/etc/init.d/boot b/package/base-files/files/etc/init.d/boot
index b36323a97e..00bae1eb53 100755
--- a/package/base-files/files/etc/init.d/boot
+++ b/package/base-files/files/etc/init.d/boot
@@ -33,6 +33,7 @@ boot() {
 	mkdir -p /tmp/resolv.conf.d
 	touch /tmp/resolv.conf.d/resolv.conf.auto
 	ln -sf /tmp/resolv.conf.d/resolv.conf.auto /tmp/resolv.conf
+	ln -sf /tmp/resolv.conf.d/resolv.conf.auto /tmp/resolv.conf.auto
 	grep -q debugfs /proc/filesystems && /bin/mount -o noatime -t debugfs debugfs /sys/kernel/debug
 	grep -q bpf /proc/filesystems && /bin/mount -o nosuid,nodev,noexec,noatime,mode=0700 -t bpf bpffs /sys/fs/bpf
 	[ "$FAILSAFE" = "true" ] && touch /tmp/.failsafe
diff --git a/package/network/services/dnsmasq/files/dhcp.conf b/package/network/services/dnsmasq/files/dhcp.conf
index 8c42ef782e..5ff9bf8bfd 100644
--- a/package/network/services/dnsmasq/files/dhcp.conf
+++ b/package/network/services/dnsmasq/files/dhcp.conf
@@ -20,6 +20,7 @@ config dnsmasq
 	#list notinterface	lo
 	#list bogusnxdomain     '64.94.110.11'
 	option localservice	1  # disable to allow DNS requests from non-local subnets
+	option filteraaaa	1
 	option ednspacket_max	1232
 
 config dhcp lan
diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
index 680e72f9e7..c998aa3ac5 100644
--- a/package/network/services/dnsmasq/files/dnsmasq.init
+++ b/package/network/services/dnsmasq/files/dnsmasq.init
@@ -893,6 +893,7 @@ dnsmasq_start()
 	append_bool "$cfg" sequential_ip "--dhcp-sequential-ip"
 	append_bool "$cfg" allservers "--all-servers"
 	append_bool "$cfg" noping "--no-ping"
+	append_bool "$cfg" filteraaaa "--filter-aaaa"
 	append_bool "$cfg" rapidcommit "--dhcp-rapid-commit"
 	append_bool "$cfg" scriptarp "--script-arp"

diff --git a/package/network/services/dnsmasq/patches/900-add-filter-aaaa-option.patch b/package/network/services/dnsmasq/patches/900-add-filter-aaaa-option.patch
new file mode 100644
index 0000000000..eaa46f82de
--- /dev/null
+++ b/package/network/services/dnsmasq/patches/900-add-filter-aaaa-option.patch
@@ -0,0 +1,59 @@
+diff -rNEZbwBdu3 a/src/dnsmasq.h b/src/dnsmasq.h
+--- a/src/dnsmasq.h	2020-04-09 00:32:53.000000000 +0800
++++ b/src/dnsmasq.h	2020-06-05 23:03:45.941176855 +0800
+@@ -269,7 +269,8 @@
+ #define OPT_IGNORE_CLID    59
+ #define OPT_SINGLE_PORT    60
+ #define OPT_LEASE_RENEW    61
+-#define OPT_LAST           62
++#define OPT_FILTER_AAAA    62
++#define OPT_LAST           63
+ 
+ #define OPTION_BITS (sizeof(unsigned int)*8)
+ #define OPTION_SIZE ( (OPT_LAST/OPTION_BITS)+((OPT_LAST%OPTION_BITS)!=0) )
+diff -rNEZbwBdu3 a/src/option.c b/src/option.c
+--- a/src/option.c	2020-04-09 00:32:53.000000000 +0800
++++ b/src/option.c	2020-06-05 23:06:29.880350271 +0800
+@@ -167,6 +167,7 @@
+ #define LOPT_IGNORE_CLID   358
+ #define LOPT_SINGLE_PORT   359
+ #define LOPT_SCRIPT_TIME   360
++#define LOPT_FILTER_AAAA   361
+  
+ #ifdef HAVE_GETOPT_LONG
+ static const struct option opts[] =  
+@@ -339,6 +340,7 @@
+     { "dumpfile", 1, 0, LOPT_DUMPFILE },
+     { "dumpmask", 1, 0, LOPT_DUMPMASK },
+     { "dhcp-ignore-clid", 0, 0,  LOPT_IGNORE_CLID },
++    { "filter-aaaa", 0, 0, LOPT_FILTER_AAAA },
+     { NULL, 0, 0, 0 }
+   };
+ 
+@@ -518,6 +520,7 @@
+   { LOPT_DUMPFILE, ARG_ONE, "<path>", gettext_noop("Path to debug packet dump file"), NULL },
+   { LOPT_DUMPMASK, ARG_ONE, "<hex>", gettext_noop("Mask which packets to dump"), NULL },
+   { LOPT_SCRIPT_TIME, OPT_LEASE_RENEW, NULL, gettext_noop("Call dhcp-script when lease expiry changes."), NULL },
++  { LOPT_FILTER_AAAA, OPT_FILTER_AAAA, NULL, gettext_noop("Filter all AAAA requests."), NULL },
+   { 0, 0, NULL, NULL, NULL }
+ }; 
+ 
+diff -rNEZbwBdu3 a/src/rfc1035.c b/src/rfc1035.c
+--- a/src/rfc1035.c	2020-04-09 00:32:53.000000000 +0800
++++ b/src/rfc1035.c	2020-06-05 23:08:46.476106541 +0800
+@@ -1955,6 +1955,15 @@
+ 	    }
+ 	}
+ 
++      //patch to filter aaaa forwards
++      if (qtype == T_AAAA && option_bool(OPT_FILTER_AAAA) ){
++        //return a null reply
++        ans = 1;
++        if (!dryrun) log_query(F_CONFIG | F_IPV6 | F_NEG, name, &addr, NULL);
++        break;
++      }
++      //end of patch
++
+       if (!ans)
+ 	return 0; /* failed to answer a question */
+     }
\ No newline at end of file
-- 
2.28.0
